In [122]:
import pandas as pd
import numpy as np
import seaborn as sns
from pydataset import data

### Exercise 1

#### Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [123]:
attendance = pd.read_csv('untidy-data/attendance.csv')
attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [124]:
attendance = attendance.rename(columns = {'Unnamed: 0':'Name'})

attendance

Name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   
2  Billy          A          T          A          A          H          T   
3   John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [125]:
attendance = attendance.melt(id_vars = 'Name',
                            var_name = 'Date',
                            value_name = 'Attendance')
attendance.head()

Name        Date Attendance
0  Sally  2018-01-01          P
1   Jane  2018-01-01          A
2  Billy  2018-01-01          A
3   John  2018-01-01          P
4  Sally  2018-01-02          T

In [126]:
attendance['Attendance'].replace({'P': 1, 'A': 0, 'T': .9, 'H': .5}, inplace = True)
attendance.head()

Name        Date  Attendance
0  Sally  2018-01-01         1.0
1   Jane  2018-01-01         0.0
2  Billy  2018-01-01         0.0
3   John  2018-01-01         1.0
4  Sally  2018-01-02         0.9

In [127]:
attendance_tidy = attendance.pivot_table(index = 'Name',
                                        columns = 'Date',
                                        values = 'Attendance')
attendance_tidy

Date   2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  2018-01-06  \
Name                                                                            
Billy         0.0         0.9         0.0         0.0         0.5         0.9   
Jane          0.0         1.0         0.9         0.9         0.9         0.9   
John          1.0         0.9         0.5         1.0         1.0         0.9   
Sally         1.0         0.9         0.9         0.5         1.0         0.0   

Date   2018-01-07  2018-01-08  
Name                           
Billy         1.0         0.9  
Jane          0.0         0.9  
John          1.0         1.0  
Sally         0.9         0.9

In [128]:
attendance_tidy['average_score'] = attendance_tidy.mean(numeric_only=True, axis=1)
attendance_tidy

Date   2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  2018-01-06  \
Name                                                                            
Billy         0.0         0.9         0.0         0.0         0.5         0.9   
Jane          0.0         1.0         0.9         0.9         0.9         0.9   
John          1.0         0.9         0.5         1.0         1.0         0.9   
Sally         1.0         0.9         0.9         0.5         1.0         0.0   

Date   2018-01-07  2018-01-08  average_score  
Name                                          
Billy         1.0         0.9         0.5250  
Jane          0.0         0.9         0.6875  
John          1.0         1.0         0.9125  
Sally         0.9         0.9         0.7625

In [129]:
attendance_tidy['average_score']

Name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: average_score, dtype: float64

### Coffee Levels

#### Read the coffee_levels.csv file.

In [130]:
df = pd.read_csv('untidy-data/coffee_levels.csv')
df

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

#### Transform the data so that each carafe is in it's own column.

In [131]:
df = df.pivot_table(index = 'hour',
              columns = 'coffee_carafe',
              values = 'coffee_amount').reset_index()
df

coffee_carafe  hour         x         y         z
0                 8  0.816164  0.189297  0.999264
1                 9  0.451018  0.521502  0.915990
2                10  0.843279  0.023163  0.144928
3                11  0.335533  0.235529  0.311495
4                12  0.898291  0.017009  0.771947
5                13  0.310711  0.997464  0.398520
6                14  0.507288  0.058361  0.864464
7                15  0.215043  0.144644  0.436364
8                16  0.183891  0.544676  0.280621
9                17  0.391560  0.594126  0.436677

### Cake Recipes

#### Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [215]:
df = pd.read_csv('untidy-data/cake_recipes.csv')
df

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

#### Tidy the data as necessary.

In [216]:
df = df.melt(id_vars = 'recipe:position',
            var_name = 'temp',
            value_name = 'score')
df

recipe:position temp      score
0         a:bottom  225  61.738655
1            a:top  225  51.709751
2         b:bottom  225  57.095320
3            b:top  225  82.455004
4         c:bottom  225  96.470207
5            c:top  225  71.306308
6         d:bottom  225  52.799753
7            d:top  225  96.873178
8         a:bottom  250  53.912627
9            a:top  250  52.009735
10        b:bottom  250  61.904369
11           b:top  250  95.224151
12        c:bottom  250  52.001358
13           c:top  250  82.795477
14        d:bottom  250  58.670419
15           d:top  250  76.101363
16        a:bottom  275  74.414730
17           a:top  275  68.576858
18        b:bottom  275  61.196980
19           b:top  275  98.594881
20        c:bottom  275  92.893227
21           c:top  275  92.098049
22        d:bottom  275  51.747686
23           d:top  275  59.571620
24        a:bottom  300  98.786784
25           a:top  300  50.228470
26        b:bottom  300  99.248541
27           b:top  300  58.169349
28        c:bottom  300  65.473084
29           c:top  300  53.960273
30        d:bottom  300  56.183110
31           d:top  300  50.971626

In [217]:
df = df.rename(columns = {'recipe:position':'recipe'})
df.columns

Index(['recipe', 'temp', 'score'], dtype='object')

In [218]:
df.head()

recipe temp      score
0  a:bottom  225  61.738655
1     a:top  225  51.709751
2  b:bottom  225  57.095320
3     b:top  225  82.455004
4  c:bottom  225  96.470207

In [219]:
df[['recipe','position']] = df.recipe.str.split(':', expand = True)
df.head()

recipe temp      score position
0      a  225  61.738655   bottom
1      a  225  51.709751      top
2      b  225  57.095320   bottom
3      b  225  82.455004      top
4      c  225  96.470207   bottom

In [220]:
df_recipes = df.pivot_table(index = 'recipe',
              columns = 'temp',
              values = 'score')

df_recipes

temp          225        250        275        300
recipe                                            
a       56.724203  52.961181  71.495794  74.507627
b       69.775162  78.564260  79.895930  78.708945
c       83.888258  67.398417  92.495638  59.716678
d       74.836465  67.385891  55.659653  53.577368

#### Which recipe, on average, is the best? recipe b

In [221]:
df_recipes = df_recipes.T
df_recipes

recipe          a          b          c          d
temp                                              
225     56.724203  69.775162  83.888258  74.836465
250     52.961181  78.564260  67.398417  67.385891
275     71.495794  79.895930  92.495638  55.659653
300     74.507627  78.708945  59.716678  53.577368

In [222]:
df_recipes.mean(axis = 0).sort_values(ascending = False)

recipe
b    76.736074
c    75.874748
a    63.922201
d    62.864844
dtype: float64

#### Which oven temperature, on average, produces the best results? 275

In [223]:
df_recipes

recipe          a          b          c          d
temp                                              
225     56.724203  69.775162  83.888258  74.836465
250     52.961181  78.564260  67.398417  67.385891
275     71.495794  79.895930  92.495638  55.659653
300     74.507627  78.708945  59.716678  53.577368

In [224]:
df_recipes['average_temp_score'] = df_recipes.mean(axis=1)
df_recipes['average_temp_score'].sort_values(ascending = False)

temp
275    74.886754
225    71.306022
300    66.627655
250    66.577437
Name: average_temp_score, dtype: float64

#### Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [234]:
df

recipe temp      score position
0       a  225  61.738655   bottom
1       a  225  51.709751      top
2       b  225  57.095320   bottom
3       b  225  82.455004      top
4       c  225  96.470207   bottom
5       c  225  71.306308      top
6       d  225  52.799753   bottom
7       d  225  96.873178      top
8       a  250  53.912627   bottom
9       a  250  52.009735      top
10      b  250  61.904369   bottom
11      b  250  95.224151      top
12      c  250  52.001358   bottom
13      c  250  82.795477      top
14      d  250  58.670419   bottom
15      d  250  76.101363      top
16      a  275  74.414730   bottom
17      a  275  68.576858      top
18      b  275  61.196980   bottom
19      b  275  98.594881      top
20      c  275  92.893227   bottom
21      c  275  92.098049      top
22      d  275  51.747686   bottom
23      d  275  59.571620      top
24      a  300  98.786784   bottom
25      a  300  50.228470      top
26      b  300  99.248541   bottom
27      b  300  58.169349      top
28      c  300  65.473084   bottom
29      c  300  53.960273      top
30      d  300  56.183110   bottom
31      d  300  50.971626      top

In [237]:
df[df.score == df.score.max()]

recipe temp      score position
26      b  300  99.248541   bottom